packages/features/bookings/Booker/utils/isTimeslotAvailable.ts

```diff
@@ -1,6 +1,68 @@
+import dayjs from "@calcom/dayjs";
+
 import type { QuickAvailabilityCheck } from "../components/hooks/useSlots";
 import { isSlotEquivalent, isValidISOFormat } from "./isSlotEquivalent";
 
+type Maybe<T> = T | undefined;
+
+// Format is YYYY-MM-DD
+type DateInBookerTimeZone = string;
+
+// Format is YYYY-MM-DDTHH:mm:ssZ
+type SlotInIsoFormat = string;
+type SlotsInIso = { time: SlotInIsoFormat }[];
+type ScheduleData = {
+  /**
+   * `slots` is a map of date in Booker's timezone to an array of time slots in ISO format
+   * So, `DateInBookerTimeZone` could be on a different day number than the slot date in ISO format
+   * For example, if Booker's timezone is UTC+5:30, and timeslot is 2025-01-02T21:00:00Z, then `DateInBookerTimeZone` could be 2025-01-03 as it is 02:30 AM next day in IST
+   */
+  slots: Record<DateInBookerTimeZone, SlotsInIso>;
+};
+
+function _isSlotPresent(slotsInIsoForDate: Maybe<SlotsInIso>, slotToCheckInIso: SlotInIsoFormat) {
+  if (!slotsInIsoForDate) return false;
+  // Check if the exact time slot exists in the available slots
+  return slotsInIsoForDate.some((slot) => {
+    return isSlotEquivalent({ slotTimeInIso: slot.time, slotToCheckInIso });
+  });
+}
+
+function _isSlotPresentInSchedule({
+  scheduleData,
+  dateInGMT,
+  slotToCheckInIso,
+}: {
+  scheduleData: ScheduleData;
+  /**
+   * It is just the date in format YYYY-MM-DD
+   */
+  dateInGMT: string;
+  slotToCheckInIso: SlotInIsoFormat;
+}) {
+  // Check if the slot is present under the date, previous date and next date
+  // Timezones can't be more than 1 day apart, so we can safely assume that previous and next day covers all the dates where a time slot can be present
+  // We don't want to look up for a slot over all the dates in the schedule data unnecessarily for performance reasons
+  const dateBefore = dayjs(dateInGMT).subtract(1, "day").format("YYYY-MM-DD");
+  const dateAfter = dayjs(dateInGMT).add(1, "day").format("YYYY-MM-DD");
```

Comment:
Using this approach of checking on previous and next date as well of a particular date to find the timeslot.

Avoiding the approach of using the booker timezone to obtain the date to lookup for as that could still be buggy if user changes timezone on client and the timezone and the currently available scheduleData is out of sync(because getSchedule request could be in progress)